### DISCUSSION TITLE
Windows 95 in DOSEMU2

### DISCUSSION THREAD

**Discussion by jschwartzenberg** [dosemu2/dosemu2#2251] [General]
![Screenshot_20240731_212930](https://github.com/user-attachments/assets/6945f39d-4cb8-4de0-a7e5-805b3023c742)

No idea why it's referred as Windows 95 as it's supposedly older than some of the Windows 3.x releases, but it runs in DOSEMU2!
More information: https://betawiki.net/wiki/Windows_95_20_July_1992_build

Unrelated, but the recent dump that people are diving into seems to have some builds that are not on that betawiki yet:
https://archive.org/details/ibm-wgam-wbiz-collection

**Comment by stsp** [comment]
> Unlike later builds of Windows 95, it requires a version of [MS-DOS](https://betawiki.net/wiki/MS-DOS)
> to be already installed and does not include any Win32 support, notably predating the earliest publicly released Win32s beta.

So while really a nice finding, this is
nothing more than win3.1 on some
steroids.

**Comment by stsp** [comment]
I really don't know why do we need win95.
If we could have directX then yes.
But otherwise its basically a waste of time.

**Reply by jschwartzenberg** [reply]
Wasn't it the same with Windows 3.1 instead of WinOS/2?
I wouldn't be surprised if there be would some business that would benefit from run their application in such an environment.

It would put the spotlight on DOSEMU2 ;)

Btw. I looked a bit through the Merge/Win4Lin data and it appears they indeed have some quite specific Windows code in there.
[dosexec.gz](https://github.com/user-attachments/files/16446659/dosexec.gz)
[em.rom.gz](https://github.com/user-attachments/files/16446661/em.rom.gz)
[emdbg.rom.gz](https://github.com/user-attachments/files/16446662/emdbg.rom.gz)

**Reply by stsp** [reply]
> Wasn't it the same with Windows 3.1 instead of WinOS/2?

2 major differences: win3.1 required
relatively few efforts, compared to win95.
win95 w/o directX is a waste of time,
whereas win3.1 support is nearly full
(in std mode at least - even DOS prompt
works). I really don't know what value
does win95 brings w/o directX (in which
case you could run Mafia or GTA3).

Does win4lin support directX?

**Reply by jschwartzenberg** [reply]
Win4Lin did not support DirectX, at least that's what they advertised. In practice many games indeed refused to run, but some worked quite well. I guess they had a full software fallback and because of the high performance that wasn't a problem.

Maybe it makes more sense to support win32s?

Certainly until the DOSBox-X people figure it all out here: https://github.com/joncampbell123/dosbox-x/issues/1217

**Reply by stsp** [reply]
> Maybe it makes more sense to support win32s?

win32s without winG also brings
you nearly nothing.

> Certainly until the DOSBox-X people figure it all out here

What they figure is irrelevant as
long as we don't have GEMMIS support.
There are 2 ways of running win95:
either you reimplement its ring0 kernel
as win4lin did, or you implement GEMMIS
and try to do it dosbox-X's way by running
win95's native ring0 kernel. This is easier
than win4lin's, but for dosemu2 that means
kvm-only (we don't have softMMU mode),
and also, as you can see, dosbox-X didn't
succeed yet. And I don't expect they will:
they currently fight with device driver header
for IFSHLP, but reading about "winbolt"
project reveals there are years of work
ahead of that. So its not even clear if this
route is actually simpler than win4lin's.
Maybe if we hire Matthias Paul and Andrew
Schulman, we can eventually get that done. :)

**Reply by stsp** [reply]
To put it straight: the cost of supporting
win95 under dosemu2 lays somewhere
between perhaps 10K$ and 40K$. And
the profit is 0 as wine already does the
job, covering directX and all the rest.

**Reply by jschwartzenberg** [reply]
> win32s without winG also brings

But WinG seems to be working already. I managed to run multiple WinG applications. What do you think is missing?


> but reading about "winbolt"
> project reveals there are years of work
> ahead of that.

Yep, exactly it would be interesting if either winbolt or win4lin sources would suddenly pop up, but without that it seems insane.


> To put it straight: the cost of supporting
> win95 under dosemu2 lays somewhere
> between perhaps 10K$ and 40K$.

It would be much more reasonable to prepare an image with a ready to run Win4Lin (Terminal Server) environment for QEMU/KVM in that case. Maybe I will look into that.

**Reply by stsp** [reply]
> But WinG seems to be working already. I managed to run multiple WinG applications. What do you think is missing?

Have you tried winG port of doom?
Ok, so what does win32s add to the
list of winG games?

> It would be much more reasonable to prepare an image with a ready to run Win4Lin (Terminal Server)
> environment for QEMU/KVM in that case.

I am not actually sure what are you up to.
Wanna play windows games under VM?
No problems: use qemu and samba for
host FS access. You'll get the guest drivers
for virtio-GPU (so directX). Wanna completely
open-source solution? Use wine.
dosemu2 simply doesn't suit: you don't
get the open-source solution (still the
win95 parts, no matter what route you
take), you don't get latest solution
(only win9x at best), you don't get directX.
So what does it give you?
Same with win4lin. I simply don't understand.
Want open-sourced win4lin alternative?
Hire me and I'll make it for you, even if I
don't understand why anyone wants that. :)

**Reply by jschwartzenberg** [reply]
> Have you tried winG port of doom?

From what I see it requires win32s.


> Ok, so what does win32s add to the
> list of winG games?

Doom for starters :)
But quite a few games listing Windows 3.1 compatibility kind of sneaked win32s along. Some that I know of are Lemmings and The Incredible Machine 3.


> I am not actually sure what are you up to.

Playing with classic software ;)

**Reply by stsp** [reply]
Ok, win32s might be nice to have.
But in a mean time you can try
runnign win32s doom with HX.
Quake2 works.

**Reply by jschwartzenberg** [reply]
The first try gives me this:
![Screenshot_20240806_164202](https://github.com/user-attachments/assets/0d8bad4f-6e7e-47b8-9f47-2e6f58a470d0)

Also I found this in the README:
```
  The HX DOS extender runtime allows to run HX DOS extended or Win32
  console applications in DOS. Furthermore there is limited support
  for running Win32 graphics applications (DirectDraw, GDI, OpenGL).
```
I suspect WinG is not supported.

**Reply by Torinde** [reply]
https://github.com/Baron-von-Riedesel/HX/issues/45

**Comment by stuaxo** [comment]
Apropos nothing, but as time passes I find it harder to Google Winbolt and WinGlue, there's a nice thread here:

https://msfn.org/board/topic/109018-windows-98-in-dr-dos/?tab=comments#comment-721209

It's a shame the TSR doesn't turn up, I bet somebody has it on some old floppy somewhere.

**Reply by stsp** [reply]
Even if you have it (w/o sources), all
you could is to "somewhat" run win95
under some version of dr-dos.
I don't think that solves any problem.
Maybe if you got winbolt sources, you
could get a bit further than that. :)

**Reply by jschwartzenberg** [reply]
> It's a shame the TSR doesn't turn up, I bet somebody has it on some old floppy somewhere.

I think he basically suggests there that he is running it himself in that thread. When I searched a bit around, I found somebody with a matching name teaching an operating system class in a German university. I wouldn't be surprised if that'd be him :)

**Reply by stsp** [reply]
I hope he doesn't teach win95 or DOS
to students. :)

**Reply by stuaxo** [reply]
Maybe it's far enough in the past for him to surface it.   I doubt MS is going to be sending some sort of lawyer ninjas at this point if it did.

**Reply by jschwartzenberg** [reply]
> Maybe it's far enough in the past for him to surface it. I doubt MS is going to be sending some sort of lawyer ninjas at this point if it did.

Reading that post carefully, I think he is suggesting that the ownership of DR-DOS is what is keeping the various developments from being merged.

**Reply by stuaxo** [reply]
Interesting DrDOS Inc (Device Logistics) was founded by Bryan Sparks (previously of Caldera)

https://en.wikipedia.org/wiki/DR-DOS#Later_versions

https://en.wikipedia.org/wiki/DeviceLogics

https://www.linkedin.com/in/bryan-sparks-a6a657253/

It was Bryan that clarified the licensing on CPM a couple of years ago

https://www.hackster.io/news/drdos-bryan-sparks-clarifies-vintage-os-cp-m-s-status-as-open-to-all-to-distribute-modify-5a410f94a8a6

I'd guess that ownership issues aren't a problem by now, though he might be able to clarify as he did with CPM if that's the case.

**Reply by stuaxo** [reply]
OK, I sent a mail to an address on a very old usenet post, this might need a bit more work to find something current - will see what happens.

**Reply by Torinde** [reply]
At the MSFN thread Matthias Paul says he hopes for:
> owner of the DR-DOS assets may decide that its commercial life is finally over and that it is due to open source the system.

[Which seems to be the case now](https://hg.pushbx.org/ecm/edrdos/rev/d4aed5f1b3e6), with the [license being open](https://hg.pushbx.org/ecm/edrdos/rev/8d01d7fe3cc1).

Related:
- https://github.com/ecm-pushbx/8086tiny/issues/3
- https://github.com/SvarDOS/edrdos

I wish you luck in getting the sources or binaries of WinBolt/WinGlue!

**Reply by stsp** [reply]
I think Matthias Paul was already spammed
enough by e-mails about that. I'd suggest not
to contact him for anyone except those directly
involved in edr-dos development. Definitely
dosemu2 is not involved into that.
